Add the swappiness tuning option to docker.

ktraghavendra opened the pull request:
Swappiness option takes 0-100, and helps to tune swappiness
behavior per container.

Signed-off-by: Raghavendra K T raghavendra.kt@linux.vnet.ibm.com

hqhq commented on docs/man/docker-create.1.md:
```diff
@@ -48,6 +48,7 @@ docker-create - Create a new container
 [**--read-only**[=*false*]]
 [**--restart**[=*RESTART*]]
 [**--security-opt**[=*[]*]]
+[**--swappiness**[=*60*]]
```
Maybe `--mem-swappiness` is a better name? Most users may confused by swap what?

ktraghavendra replied:
I agree, will change the name

ktraghavendra replied:
Done the change to mem-swappiness

ktraghavendra committed:
Add the swappiness tuning option to docker.
    
Swappiness option takes 0-100, and helps to tune swappiness
behavior per container.

Signed-off-by: Raghavendra K T <raghavendra.kt@linux.vnet.ibm.com>

thaJeztah commented on api/types/types.go:
```diff
@@ -150,6 +150,7 @@ type Info struct {
 	Debug              bool
 	NFd                int
 	OomKillDisable     bool
+	MemSwappiness      bool
```
Should this be `int`?

ktraghavendra replied:
You are right.. I think then I will also have to use it in api/types/info.go.. Do you think it is important information to be displayed?.

ktraghavendra replied:
Sorry I meant api/client/info.go

ktraghavendra replied:
I think we do not need that for client itself.. will check that

ktraghavendra replied:
@thaJeztah , Sorry for confusing comment, so what I meant was api/info was intended for systemwide info and 
the current mem-swappiness is per container I am removing above info from api/types/types.go.
Let me know if that is okay

unclejack commented:
Is this actually namespaced?

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -61,6 +61,15 @@ func (s *DockerSuite) TestRunWithoutMemoryswapLimit(c *check.C) {
 }
 
+func (s *DockerSuite) TestRunWithSwappiness(c *check.C) {
+	testRequires(c, NativeExecDriver)
```
Why does this require native? It looks like you have this in the LXC template.

Also, can we at least check the hostconfig output to make sure it actually got set?  Is there some other value we can verify somewhere?

ktraghavendra replied:
This is per container setting and I have verified that after creating container with mem-swappiness 0 for example the value in /sys/fs/cgroup/memory/system.slice/docker-xxx/memory.swappiness changes as expected

ktraghavendra replied:
Brian, 
>Why does this require native? It looks like you have this in the LXC template.
could you let me know what is expected from me here.. will change accordingly..

cpuguy83 replied:
This line here would make this test be skipped when running with the LXC driver.
Since it seems like this should work just fine on LXC.... so can probably just remove the line.

ktraghavendra replied:
@cpuguy83 Okay Got it .. Do not think we need that..

ktraghavendra replied:
@cpuguy83 , ofcourse corresponding libcontainer change was needed to make sure
mem-swappiness got reflected in cgroup config
https://github.com/docker/libcontainer/pull/622

cpuguy83 commented on docs/sources/reference/api/docker_remote_api_v1.20.md:
```diff
@@ -201,6 +202,7 @@ Json Parameters:
 -   **CpusetCpus** - String value containing the `cgroups CpusetCpus` to use.
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
+-   **MemSwappiness** - Integer Value, Tuning Container memory swappiness.
```
Would be helpful to put the allowed range here.

ktraghavendra replied:
Sure will update that

cpuguy83 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1020,6 +1020,7 @@ Creates a new container.
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --security-opt=[]          Security options
+      --mem-swapiness=[]         Tuning container memory swappiness
```
Also allowed range here, if it will fit.

ktraghavendra replied:
Sure..

cpuguy83 commented on pkg/sysinfo/sysinfo_linux.go:
```diff
@@ -33,6 +33,11 @@ func New(quiet bool) *SysInfo {
 		if !sysInfo.OomKillDisable && !quiet {
 			logrus.Warnf("Your kernel does not support oom control.")
 		}
+		_, err = ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.swappiness"))
```
Really not sure about all these warnings.
mem limit makes sense because it is pretty fundamental to resource control and somewhat important to know at daemon startup.

ktraghavendra replied:
@cpuguy83 since mem-swappiness is not critical shall I change such that this does not make bailing out of container creation ?
so that atleast we know that swappiness did not work from warning..?
something like (similar to swap memory):
 _, err1 = ioutil.ReadFile(path.Join(cgroupMemoryMountpoint, "memory.swappiness"))
                sysInfo.MemSwappiness = err1 == nil
                if !sysInfo.MemSwappiness && !quiet {
                        logrus.Warnf("Your kernel does not support container swappiness control.")
                }
OR should I remove whole check?

ktraghavendra replied:
@cpuguy83 , that does not solve bail out problem .. I am removing the check itself..

cpuguy83 commented:
@unclejack This is a cgroup setting, not the one from `/proc/sys/vm/swappiness`

cpuguy83 commented:
Ok, we need to close this for now since it's not mergeable as is until the corresponding change makes it into libcontainer, and the updated version vendored into Docker.

estesp commented on docs/sources/reference/run.md:
```diff
@@ -604,6 +605,17 @@ The following example, illustrates a dangerous way to use the flag:
 and require killing system processes to free memory.
 
+### Swappiness constraint
+By default, some percenatge of anonymous pages of container can be swapped out.
```
spelling: *percentage*

ktraghavendra replied:
Thanks phil done (in https://github.com/docker/docker/pull/14004)
